File: src/main/java/com/history/tree/services/PersonService.java
package com.history.tree.services;

import com.history.tree.dto.PersonDTO;
import com.history.tree.mappers.PersonMapper;
import com.history.tree.repositories.PersonRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

@Service
@RequiredArgsConstructor
public class PersonService {

    private final PersonRepository repository;

    public Mono<PersonDTO> findById(long id) {
        return repository.findById(id).map(PersonMapper.INSTANCE::personToDto);
    }

    public Flux<PersonDTO> getPersonsByTreeId(long treeId) {
        return repository.findAllByTreeId(treeId).map(PersonMapper.INSTANCE::personToDto);
    }

}

File: src/main/java/com/history/tree/services/PersonService.kt
package com.history.tree.services

import com.history.tree.dto.PersonDTO
import com.history.tree.mappers.PersonMapper
import com.history.tree.repositories.PersonRepository
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import org.mapstruct.factory.Mappers
import org.springframework.stereotype.Service

@Service
class PersonService(val repository: PersonRepository) {

    suspend fun findById(id: Long): PersonDTO {
        val person = repository.findById(id)

        val mapper = Mappers.getMapper(PersonMapper::class.java)

        return mapper.personToDto(person!!)
    }

    suspend fun getPersonsByTreeId(treeId: Long): Flow<PersonDTO> {
        val mapper = Mappers.getMapper(PersonMapper::class.java)

        return repository.findAllByTreeId(treeId)
            .map { person -> mapper.personToDto(person) }
    }
}